--- Makefile.orig	2021-04-04 12:34:57.000000000 -0400
+++ Makefile	2022-10-21 15:59:03.280208400 -0400
@@ -10,7 +10,9 @@
 SHARED_LIB = libpasswdqc.so.1
 DEVEL_LIB = libpasswdqc.so
 SHARED_LIB_DARWIN = libpasswdqc.0.dylib
+SHARED_LIB_CYGWIN = cygpasswdqc-0.dll
 DEVEL_LIB_DARWIN = libpasswdqc.dylib
+DEVEL_LIB_CYGWIN = libpasswdqc.dll.a
 MAP_LIB = libpasswdqc.map
 PAM_SO_SUFFIX =
 SHARED_PAM = $(TITLE).so$(PAM_SO_SUFFIX)
@@ -34,6 +36,7 @@
 CONFDIR = /etc
 CONFMODE = 644
 SHARED_LIBDIR = /lib
+SHARED_LIBDIR_CYGWIN = /usr/bin
 SHARED_LIBDIR_SUN = /usr/lib
 SHARED_LIBDIR_REL = ../..$(SHARED_LIBDIR)
 DEVEL_LIBDIR = /usr/lib
@@ -78,6 +81,10 @@
 	-Wl,--soname,$(SHARED_LIB),--version-script,$(MAP_LIB)
 LDFLAGS_lib_SUN = $(LDFLAGS_shared_SUN)
 LDFLAGS_lib_HP = $(LDFLAGS_shared_HP)
+LDFLAGS_lib_CYGWIN = $(LDFLAGS_shared) \
+	-Wl,--out-implib=$(DEVEL_LIB_CYGWIN) \
+	-Wl,--export-all-symbols \
+	-Wl,--enable-auto-import
 LDFLAGS_pam = $(LDFLAGS_shared)
 LDFLAGS_pam_LINUX = $(LDFLAGS_shared_LINUX) \
 	-Wl,--version-script,$(MAP_PAM)
@@ -104,6 +111,7 @@
 
 CONFIGS = passwdqc.conf
 BINS = pwqgen pwqcheck pwqfilter
+BINS_CYGWIN = $(BINS) $(SHARED_LIB_CYGWIN)
 PROJ = $(SHARED_LIB) $(DEVEL_LIB) $(SHARED_PAM) $(BINS)
 OBJS_LIB = concat.o md4.o passwdqc_check.o passwdqc_filter.o passwdqc_load.o passwdqc_memzero.o passwdqc_parse.o passwdqc_random.o wordset_4k.o
 OBJS_PAM = pam_passwdqc.o passwdqc_memzero.o
@@ -113,7 +121,7 @@
 
 default: all
 
-all locales pam utils install install_lib install_locales install_pam install_utils uninstall remove remove_lib remove_locales remove_pam remove_utils:
+all locales pam utils install install_lib install_locales install_pam install_utils uninstall remove remove_lib remove_locales remove_pam remove_utils clean:
 	case "`uname -s`" in \
 	Linux)	$(MAKE) CPPFLAGS_lib="$(CPPFLAGS_lib) -DHAVE_SHADOW" \
 			LDFLAGS_lib="$(LDFLAGS_lib_LINUX)" \
@@ -141,6 +149,14 @@
 			SECUREDIR="$(SECUREDIR_DARWIN)" \
 			LDLIBS_pam="$(LDLIBS_pam_DARWIN)" \
 			$@_wrapped;; \
+	CYGWIN_NT*)	$(MAKE) CPPFLAGS_lib="$(CPPFLAGS_lib)" \
+			SHARED_LIB="$(SHARED_LIB_CYGWIN)" \
+			SHARED_LIBDIR="$(SHARED_LIBDIR_CYGWIN)" \
+			DEVEL_LIB="$(DEVEL_LIB_CYGWIN)" \
+			LDFLAGS_lib="$(LDFLAGS_lib_CYGWIN)" \
+			BINS="$(BINS_CYGWIN)" \
+			CYGWIN=true \
+			$@_wrapped;; \
 	*)	$(MAKE) $@_wrapped;; \
 	esac
 
@@ -154,7 +170,9 @@
 	$(LD_lib) $(LDFLAGS_lib) $(OBJS_LIB) $(LDLIBS_lib) -o $(SHARED_LIB)
 
 $(DEVEL_LIB): $(SHARED_LIB)
+ifndef CYGWIN
 	$(LN_s) $(SHARED_LIB) $(DEVEL_LIB)
+endif
 
 $(SHARED_PAM): $(OBJS_PAM) $(MAP_PAM) $(DEVEL_LIB)
 	$(LD_lib) $(LDFLAGS_pam) $(OBJS_PAM) $(LDLIBS_pam) -L. -lpasswdqc -o $(SHARED_PAM)
@@ -200,8 +218,12 @@
 	$(INSTALL) -m $(SHLIBMODE) $(SHARED_LIB) $(DESTDIR)$(SHARED_LIBDIR)/
 
 	$(MKDIR) $(DESTDIR)$(DEVEL_LIBDIR)
+ifndef CYGWIN
 	$(LN_s) $(SHARED_LIBDIR_REL)/$(SHARED_LIB) \
 		$(DESTDIR)$(DEVEL_LIBDIR)/$(DEVEL_LIB)
+else
+	$(INSTALL) -m $(SHLIBMODE) $(DEVEL_LIB) $(DESTDIR)$(DEVEL_LIBDIR)/
+endif
 
 	$(MKDIR) $(DESTDIR)$(INCLUDEDIR)
 	$(INSTALL) -m $(INCMODE) $(HEADER) $(DESTDIR)$(INCLUDEDIR)/
@@ -274,7 +296,7 @@
 remove_locales_wrapped:
 	for f in $(LANGUAGES); do $(RM) $(DESTDIR)$(LOCALEDIR)/$$f/LC_MESSAGES/$(PACKAGE).mo; done
 
-clean:
+clean_wrapped:
 	$(RM) $(PROJ) $(POTFILE) $(MOFILES) *.o
 
 .PHONY: all all_wrapped clean install install_lib install_locales install_pam install_utils \
